util/dvsim/CfgFactory.py

```diff
@@ -0,0 +1,103 @@
+
+    Returns a pair (cls, hjson_data) on success or raises a RuntimeError on
+    failure.
+
+    '''
+    # Start by loading up the hjson file and any included files
+    hjson_data = load_hjson(path, initial_values)
+
+    # Look up the value of dvsim_cfg_type in the loaded data. This is a
+    # required field, and tells us what sort of FlowCfg to make.
+    cfg_type = hjson_data.get('dvsim_cfg_type')
+    if cfg_type is None:
+        raise RuntimeError('{!r}: Configuration file does not define '
+                           'dvsim_cfg_type. Are you sure this is a dvsim '
+                           'configuration file?'
+                           .format(path))
+
+    classes = [
+        LintCfg.LintCfg,
+        SynCfg.SynCfg,
+        FpvCfg.FpvCfg,
+        SimCfg.SimCfg
+    ]
+    found_cls = None
+    known_types = []
+    for cls in classes:
+        assert cls.cfg_type_key is not None
+        known_types.append(cls.cfg_type_key)
+        if cls.cfg_type_key == cfg_type:
+            found_cls = cls
+            break
+    if found_cls is None:
+        raise RuntimeError('{}: Configuration file sets dvsim_cfg_type '
+                           'to {!r}, but this is not a known type (known '
+                           'types: {}).'
+                           .format(path, cfg_type, ', '.join(known_types)))
+
+    return (found_cls, hjson_data)
+
+
+def _make_child_config(path, args, initial_values):
+    try:
+        cls, hjson_data = _load_config(path, initial_values)
+    except RuntimeError as err:
+        log.error(str(err))
+        sys.exit(1)
+
+    # Since this is a child configuration (from some primary configuration),
+    # make sure that we aren't ourselves a primary configuration. We don't need
+    # multi-level hierarchies and this avoids circular dependencies.
+    if 'use_cfgs' in hjson_data:
+        raise RuntimeError('{}: Configuration file has use_cfgs, but is '
+                           'itself included from another configuration.'
+                           .format(path))
+
+    # Call cls as a constructor. Note that we pass None as the mk_config
+    # argument: this is not supposed to load anything else.
+    return cls(path, hjson_data, args, None)
+
+
+def make_config(path, args, proj_root):
+    '''Make a flow config by loading the config file at path
+
+    args is the arguments passed to the dvsim.py tool, proj_root is the top of
+    the project and is_toplevel is true if this is the first configuration to
+    be loaded (we aren't a child of some primary config).
+
+    '''
+    initial_values = {'proj_root': proj_root}
+    if args.tool is not None:
+        initial_values['tool'] = args.tool
+
+    try:
+        cls, hjson_data = _load_config(path, initial_values)
+    except RuntimeError as err:
+        log.error(str(err))
+        sys.exit(1)
+
+    def factory(child_path):
+        child_ivs = initial_values.copy()
+        child_ivs['dvsim_cfg_type'] = hjson_data['dvsim_cfg_type']
```

Comment:
A cfgs list is just a dump list of things to iterate over. We shouldn't have to require it to specify the flow. Is there a way we could avoid that? I was envisioning in future, being able to combine heterogenous flows into the same cfgs list (which of course requires that whole logic to be refactored out).

Reply:
I agree. Probably the cleanest approach is to make a primary/cfgs_list/parent/whatever config into a separate type. We could probably get rid of the recursion here (doing the recursive loads in the "`ListCfg`" type) and it would also simplify lots of the code in the FlowCfg subclasses (which currently have to work out whether they are primary configs).

However: Doing that cleanly depends on the changes in this PR, so can we get this in first? Then one of us can do the follow-up changes afterwards.

Reply:
Yes, agree. That part needs to be cleaned up.